engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/hydration/batch/NadelBatchHydrationObjectIdFieldBuilder.kt

```diff
@@ -0,0 +1,132 @@
+ *      service: "IssueService"
+ *      field: "issues"
+ *      arguments : [{ name: "ids" value: "$source.relatedIssueIds" }]
+ *      identifiedBy: "id"
+ *      batchSize: 2
+ *   )
+ * }
+ * ```
+ *
+ * And given a hydration query
+ *
+ * ```graphql
+ * query {
+ *   issues(ids: ["PROJ-1", "PROJ-5", "PROJ-3"]) {
+ *      key
+ *   }
+ * }
+ * ```
+ *
+ * We need an id to actually identify the objects to know where to insert them back
+ * into the original result.
+ *
+ * i.e. this code inserts the selection for the `id` field
+ */
+internal object NadelBatchHydrationObjectIdFieldBuilder {
+    fun makeObjectIdField(
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        aliasHelper: NadelAliasHelper,
+        batchHydrationInstruction: NadelBatchHydrationFieldInstruction,
+    ): ExecutableNormalizedField? {
+        return when (val matchStrategy = batchHydrationInstruction.batchHydrationMatchStrategy) {
+            is NadelBatchHydrationMatchStrategy.MatchObjectIdentifier -> makeObjectIdField(
+                executionBlueprint,
+                aliasHelper,
+                batchHydrationInstruction,
+                matchStrategy,
+            )
+            else -> null
+        }
+    }
+
+    private fun makeObjectIdField(
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        aliasHelper: NadelAliasHelper,
+        batchHydrationInstruction: NadelBatchHydrationFieldInstruction,
+        matchStrategy: NadelBatchHydrationMatchStrategy.MatchObjectIdentifier,
+    ): ExecutableNormalizedField {
+        val objectTypeNames = getObjectTypeNamesForIdField(
+            executionBlueprint = executionBlueprint,
+            actorService = batchHydrationInstruction.actorService,
+            underlyingParentTypeOfIdField = batchHydrationInstruction.actorFieldDef.type,
+        )
+
+        return newNormalizedField()
+            .objectTypeNames(objectTypeNames)
+            .fieldName(matchStrategy.objectId)
+            .alias(aliasHelper.getResultKey(matchStrategy.objectId))
+            .build()
+    }
+
+    /**
+     * Gets the type names for the parent of the object id field. Note that this
+     * must be the OVERALL type names.
+     */
+    private fun getObjectTypeNamesForIdField(
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        actorService: Service,
+        underlyingParentTypeOfIdField: GraphQLOutputType,
+    ): List<String> {
+        val overallType = getUnwrappedOverallType(
+            executionBlueprint = executionBlueprint,
+            service = actorService,
+            underlyingType = underlyingParentTypeOfIdField,
+        ) ?: error("Could not find the overall output type for the actor field")
+
+        return resolveObjectTypes(
+            schema = executionBlueprint.schema,
+            type = overallType,
+            onNotObjectType = {
+                errorForUnsupportedObjectIdParentType()
+            },
+        ).map { it.name }.toList()
```

Comment:
`.toList()`
do we need it? It's already a list I think

Reply:
Ah yes you're right idk why I did that…

Reply:
Fixed, thanks